src/Juvix/Backends/LLVM/Codegen/Graph.hs

```diff
@@ -29,46 +33,58 @@ link ∷
     HasState "symtab" (Map.HashMap Symbol Operand.Operand) m
   ) ⇒
   m Operand.Operand
-link = body >>= define Type.void "link" args
+link = Block.defineFunction Type.void "link" args $
+  do
+    setPort ("node_1", "port_1") ("node_2", "port_2")
+    setPort ("node_2", "port_2") ("node_1", "port_1")
+    retNull
   where
     args =
-      ( [ (nodeType, "node_1"),
-          (numPorts, "port_1"),
-          (nodeType, "node_2"),
-          (numPorts, "port_2")
-        ] ∷
-          [(Type.Type, Name.Name)]
-      )
-    -- TODO ∷ Abstract most of the logic in this function
-    body = do
-      makeFunction "link" args
-      setPort ("node_1", "port_1") ("node_2", "port_2")
-      setPort ("node_2", "port_2") ("node_1", "port_1")
-      _ ← retNull
-      createBlocks
+      [ (nodeType, "node_1"),
+        (numPorts, "port_1"),
+        (nodeType, "node_2"),
+        (numPorts, "port_2")
+      ]
 
 -- perform offsets
 
-isBothPrimary = body >>= define Type.i1 "is_both_primary" args
+isBothPrimary ∷
+  ( HasThrow "err" Errors m,
+    HasState "blockCount" Int m,
+    HasState "blocks" (Map.HashMap Name.Name BlockState) m,
+    HasState "count" Word m,
+    HasState "currentBlock" Name.Name m,
+    HasState "moduleDefinitions" [AST.Definition] m,
+    HasState "names" Names m,
+    HasState "symtab" (Map.HashMap Symbol Operand.Operand) m,
+    HasState "typTab" TypeTable m,
+    HasState "varTab" VariantToType m
+  ) ⇒
+  m Operand.Operand
+isBothPrimary = Block.defineFunction Type.i1 "is_both_primary" args $
+  do
+    -- TODO ∷ should this call be abstracted somewhere?!
+    -- Why should Ι allocate for every port?!
+    mainPort ← allocaNumPortsStatic False (Operand.ConstantOperand (C.Int 32 0))
+    -- TODO ∷ Make sure findEdge is in the environment
+    edge ← Block.externf "find_edge"
+    nodePtr ← Block.externf "node_ptr"
+    node ← load nodeType nodePtr
```

Comment:
Are we loading the node data here? Do we need to do that to read the port address?

Reply:
we need to load the node, as find_edge requires a nodeType and not a nodePointer